custom_components/sems/sensor.py

```diff
@@ -5,38 +5,567 @@
+                        SensorStateClass.MEASUREMENT,
+                    ),
+                    SensorOptions(
+                        device_info,
+                        f"{serial_number}-{idx}-bms_discharge_i_max",
+                        f"Inverter {inverter['name']} Battery {idx} BMS Discharge Max Current",
+                        path_to_battery + ["bms_discharge_i_max"],
+                        SensorDeviceClass.CURRENT,
+                        ELECTRIC_CURRENT_AMPERE,
+                        SensorStateClass.MEASUREMENT,
+                    ),
+                    SensorOptions(
+                        device_info,
+                        f"{serial_number}-{idx}-bms_charge_i_max",
+                        f"Inverter {inverter['name']} Battery {idx} BMS Charge Max Current",
+                        path_to_battery + ["bms_charge_i_max"],
+                        SensorDeviceClass.CURRENT,
+                        ELECTRIC_CURRENT_AMPERE,
+                        SensorStateClass.MEASUREMENT,
+                    ),
+                ]
+
+    if "hasPowerflow" in data and data["hasPowerflow"] and "powerflow" in data:
+        inverter_serial_number = get_home_kit_sn(data)
+        if not has_existing_homekit_entity or inverter_serial_number is None:
+            inverter_serial_number = "powerflow"
+        serial_backwards_compatibility = (
+            "homeKit"  # the old code uses homeKit for the serial number
+        )
+        device_info = DeviceInfo(
+            identifiers={
+                # Serial numbers are unique identifiers within a specific domain
+                (DOMAIN, serial_backwards_compatibility)
+            },
+            name="HomeKit",
+            manufacturer="GoodWe",
+        )
+        sensors += [
+            SensorOptions(
+                device_info,
+                f"{inverter_serial_number}",  # backwards compatibility otherwise would be f"{serial_number}-load"
+                f"HomeKit Load",
+                ["powerflow", "load"],
+                SensorDeviceClass.POWER,
+                POWER_WATT,
+                SensorStateClass.MEASUREMENT,
+            ),
+            SensorOptions(
+                device_info,
+                f"{inverter_serial_number}-pv",
+                f"HomeKit PV",
+                ["powerflow", "pv"],
+                SensorDeviceClass.POWER,
+                POWER_WATT,
+                SensorStateClass.MEASUREMENT,
+            ),
+            SensorOptions(
+                device_info,
+                f"{inverter_serial_number}-grid",
+                f"HomeKit Grid",
+                ["powerflow", "grid"],
+                SensorDeviceClass.POWER,
+                POWER_WATT,
+                SensorStateClass.MEASUREMENT,
+            ),
+            SensorOptions(
+                device_info,
+                f"{inverter_serial_number}-battery",
+                f"HomeKit Battery",
+                ["powerflow", GOODWE_SPELLING.battery],
```

Comment:
betteryStatus should be used to determine charging/discharging

Reply:
Do you know what the possible values of `betteryStatus` are and what they mean?

Reply:
-1 = Discharging
0 = Off (100% soc and PV production for example)
1 = Charging

Reply:
So we do
```python
batteryStatus = 0
battery = 10 // kwh
return battery * batteryStatus  // 0
```

```python
batteryStatus = 1
battery = 10 // kwh
return battery * batteryStatus  // 10
```

```python
batteryStatus = -1
battery = 10 // kwh
return battery * batteryStatus  // -10
```

if we do the same for grid it should also work for you where it's always 1 but your grid value includes the sign
e.g.
```python
gridStatus = 1
grid= -10 // kwh
return battery * batteryStatus  // -10
```